Added IAM Role Authentication using Web Identity Token

maliankurh opened the pull request.

dekobon commented:
Hi @maliankurh, are you willing / interested on further refining this PR? If not, my apologies for saying "do a PR" because I just realized that I can't contribute to the branch that you did the PR on.

mvanholsteijn commented on common/etc/nginx/include/s3gateway.js:
```diff
@@ -994,6 +1003,38 @@ async function _fetchEC2RoleCredentials() {
+    
+    var params = "Version=2011-06-15&Action=AssumeRoleWithWebIdentity&RoleArn=" + arn + "&RoleSessionName=" + name + "&WebIdentityToken=" + token;
+
+    var response = await ngx.fetch("https://sts.us-east-1.amazonaws.com?" + params, {
```
_By default, the AWS Security Token Service (AWS STS) is available as a global service, and all STS requests go to a single endpoint at https://sts.amazonaws.com/. AWS recommends using Regional STS endpoints to reduce latency, build in redundancy, and increase session token validity._

https://docs.aws.amazon.com/general/latest/gr/sts.html#sts_region

I would recommend adding an environment AWS_STS_ENDPOINT to configure the desired endpoint, while defaulting to sts.amazonaws.com.

maliankur replied:
Hello @mvanholsteijn, sure I will make the changes accordingly.

dekobon replied:
I agree with the above assessment.

maliankur commented:
> Hi @maliankurh, are you willing / interested on further refining this PR? If not, my apologies for saying "do a PR" because I just realized that I can't contribute to the branch that you did the PR on.

Hello @dekobon, Sure Let me know what changes you want me to make for the PR.

mvanholsteijn commented:
@dekobon can you specify what changes you need, so that this PR can go through?

dekobon commented on common/etc/nginx/include/s3gateway.js:
```diff
@@ -910,6 +910,15 @@ async function fetchCredentials(r) {
             return;
         }
+    }
+    else if(process.env['AWS_WEB_IDENTITY_TOKEN_FILE']) {
```
Please document this environment variable in the documentation.

dekobon commented on common/etc/nginx/include/s3gateway.js:
```diff
@@ -994,6 +1003,38 @@ async function _fetchEC2RoleCredentials() {
+ */
+async function _fetchWebIdentityCredentials(r) {
+    var arn   = process.env['AWS_ROLE_ARN'];
+    var name  = process.env['HOSTNAME'] || 'nginx-s3-gateway';
```
Please document all new environment variables.

dekobon reviewed (changes requested):
In addition to the requested changes, please add:
* Automated tests that validate the configuration (if possible)
* Use case documentation that describes what you would accomplish and how you would configure the gateway

Also, please test the other host credential based access configurations in order to see if this change introduces a regression.

Thank you so much for your contribution!

maliankurh commented:
@dekobon, I have updated the Documentation and added the steps for the STS to work. Please check and let me know if you want me to update anything else.

@mvanholsteijn, I have added Env Variable for STS Endpoint and set the default value as well. Please check

dekobon commented:
It seems that the automated test run is failing. Can you look into why the tests are failing: https://github.com/nginxinc/nginx-s3-gateway/runs/8277852689?check_suite_focus=true

maliankurh commented:
Yes I looked into it and the latest commit should fix the issue.

dekobon commented:
Awesome, let's rebase the latest changes from master into this branch and I will try to do some testing with the branch myself. Let's aim for a merge next week.

If anyone else who is following the PR would like to do their own testing, that would be much appreciated.

dekobon commented:
I want to set expectations for everyone clearly. I've been unable to do the testing that I wanted to this week. I'll aim to be working on it starting on 9/19. My apologies for the delay folks.

dm3ch commented:
I just tested this PR in my own EKS, cause I have similar case.

What I've found:
- It's not working for me without https://github.com/nginxinc/nginx-s3-gateway/pull/52/files#diff-3077c17ca3f4dd131f26585ea1026611c248ff1b3268e70c17a33b35607e2d58R82
- I'm not sure why (maybe cause of version upgrade in master or different credentials) but for me `PROVIDE_INDEX_PAGE` stoped working

dekobon commented:
I just tried to run: `./test.sh oss` using this branch and the test script returned the following failure:
```
...
▶ Building Docker Compose environment
[+] Running 3/3
 ⠿ Container ngt-minio-1             Created                                                                       0.8s
 ⠿ Container ngt-minio-client-1      Created                                                                       0.6s
 ⠿ Container ngt-nginx-s3-gateway-1  Created                                                                       0.8s
▶ Adding test data to container
Copying contents of /home/elijah/Development/F5/nginx-s3-gateway/test/data to Docker container ngt_minio_1:/
Error: No such container:path: ngt_minio_1:/
Error running tests - outputting container logs
```

I'm going to investigate the cause and update this discussion.

dekobon commented:
After going through the differences between this branch and master, it seems like it is a number of commits behind master. I think it would be best to create a new branch under the main project with the contributions in this PR, merge in the changes from master, and then try to pull in as many volunteers into testing it as possible.

dekobon commented:
I've created a new branch: https://github.com/nginxinc/nginx-s3-gateway/tree/issue/40
And a new PR for integrating that branch into master: https://github.com/nginxinc/nginx-s3-gateway/pull/54

If anyone has fixes or improvements, please do a PR against the [issue/40](https://github.com/nginxinc/nginx-s3-gateway/tree/issue/40) branch.

In this new branch, I've merged in the latest changes from master. After doing that, the automated tests now pass.

I'm going to close this PR and have us move our development to the other PR. I hope this isn't too confusing. I'm eager to help and see this feature added.
